lib/bump/dependency_file_updaters/python.rb

```diff
@@ -32,27 +33,29 @@ def updated_requirements_file
       def updated_requirements_content
         return @updated_requirements_content if @updated_requirements_content
 
-        packages = @packages.map do |pkg|
-          next pkg unless pkg.name == dependency.name
-          next pkg unless pkg.version
-          old_version = pkg.version
+        requirements.content.
+          to_enum(:scan,
+                  DependencyFileParsers::Python::LineParser::REQUIREMENT_LINE).
+          find { Regexp.last_match[:name] == dependency.name }
 
-          Dependency.new(
-            name: pkg.name,
-            version: updated_version_string(old_version, dependency.version)
-          )
-        end
+        original_dep_declaration_string = Regexp.last_match.to_s
+        updated_dep_declaration_string =
+          original_dep_declaration_string.
+          sub(DependencyFileParsers::Python::LineParser::REQUIREMENT) do |old|
+            old_version =
+              old.match(DependencyFileParsers::Python::LineParser::VERSION)[0]
 
-        @updated_requirements_content = packages.map do |pkg|
-          "#{pkg.name}==#{pkg.version}"
-        end.join("\n") + "\n"
-      end
+            precision = old_version.split(".").count
+            new_version =
+              dependency.version.split(".").first(precision).join(".")
 
-      def updated_version_string(old_version_string, new_version_number)
-        old_version_string.sub(/[\d\.]*\d/) do |old_version_number|
-          precision = old_version_number.split(".").count
-          new_version_number.split(".").first(precision).join(".")
-        end
+            old.sub(old_version, new_version)
+          end
```

Comment:
I see why this is necessary, and we do similar stuff for Ruby, but it's nasty nonetheless. We're storing state in global variables and using regexes that are internal to the parser, and doing some quite implicit stuff.

I think you could probably do this differently by pulling out more data at the parsing stage, and then using that to reconstruct the line.

e.g.
```
luigi[foo]==2.0.1 # a comment
```
becomes
```ruby
{
  name: "luigi",
  extras: ["foo"],
  requirements: [
    { comparison: "==", version: "2.0.1" }
  ],
  comment: "a comment"
}
```
and then you bump it
```ruby
{
  name: "luigi",
  extras: ["foo"],
  requirements: [
    { comparison: "==", version: "2.1.0" }
  ],
  comment: "a comment"
}
```
and reconstruct the line by following a standard format like
```ruby
extras_part = extras.any? ? "[#{extras.join(", ")}]" : ""
"#{name}#{extras_part}#{requirements.map { |r| r[:comparison]+r[:version] }.join(", ")} # #{comment}"
```

It's probably not worth doing now - I'm more just thinking aloud - but it'd make a nice future refactor.